subprojects/model-core/src/main/java/org/gradle/internal/properties/annotations/AbstractTypeMetadataWalker.java

```diff
@@ -38,47 +45,58 @@ private AbstractTypeMetadataWalker(TypeMetadataStore typeMetadataStore, Class<?
 
     @Override
     public void walk(T root, NodeMetadataVisitor<T> visitor) {
-        walk(root, null, visitor);
+        walk(root, null, visitor, emptySet());
     }
 
-    private void walk(T node, @Nullable String qualifiedName, NodeMetadataVisitor<T> visitor) {
+    private void walk(T node, @Nullable String qualifiedName, NodeMetadataVisitor<T> visitor, Set<T> previousNestedNodesWalkedOnPath) {
         Class<?> nodeType = resolveType(node);
         TypeMetadata typeMetadata = typeMetadataStore.getTypeMetadata(nodeType);
         if (Provider.class.isAssignableFrom(nodeType)) {
-            handleProvider(node, child -> walk(child, qualifiedName, visitor));
+            handleProvider(node, child -> walk(child, qualifiedName, visitor, previousNestedNodesWalkedOnPath));
         } else if (Map.class.isAssignableFrom(nodeType) && !typeMetadata.hasAnnotatedProperties()) {
-            handleMap(node, (name, child) -> walk(child, getQualifiedName(qualifiedName, name), visitor));
+            handleMap(node, (name, child) -> walk(child, getQualifiedName(qualifiedName, name), visitor, previousNestedNodesWalkedOnPath));
         } else if (Iterable.class.isAssignableFrom(nodeType) && !typeMetadata.hasAnnotatedProperties()) {
-            handleIterable(node, (name, child) -> walk(child, getQualifiedName(qualifiedName, name), visitor));
+            handleIterable(node, (name, child) -> walk(child, getQualifiedName(qualifiedName, name), visitor, previousNestedNodesWalkedOnPath));
         } else {
-            visitor.visitNested(typeMetadata, qualifiedName, node);
-            typeMetadata.getPropertiesMetadata().forEach(propertyMetadata -> {
-                String childQualifiedName = getQualifiedName(qualifiedName, propertyMetadata.getPropertyName());
-                if (propertyMetadata.getPropertyType() == nestedAnnotation) {
-                    T child = getChild(node, propertyMetadata);
-                    walk(child, childQualifiedName, visitor);
-                } else {
-                    visitor.visitLeaf(childQualifiedName, propertyMetadata, () -> getChild(node, propertyMetadata));
-                }
-            });
+            handleNested(node, typeMetadata, qualifiedName, visitor, previousNestedNodesWalkedOnPath);
         }
     }
 
-    private static String getQualifiedName(@Nullable String parentPropertyName, String childPropertyName) {
-        return parentPropertyName == null
-            ? childPropertyName
-            : parentPropertyName + "." + childPropertyName;
+    private void handleNested(T node, TypeMetadata typeMetadata, @Nullable String qualifiedName, NodeMetadataVisitor<T> visitor, Set<T> previousNestedNodesOnPath) {
+        if (previousNestedNodesOnPath.contains(node)) {
+            return;
```

Comment:
This works for the static case, but for the instance case we need to throw here.

When we throw, it would be good to show an error message with a list of the nodes in the cycle. In the message we need to list the nodes in order, and get the names of the properties themselves. The currently used `IdentityHashMap` does not provide either of these.

We could use a `LinkedHashSet<NamedValue>` where `NamedValue` would only take into account the identity of the value itself for equality.

Reply:
Now we throw exception.